### PR TITLE
Remove Engine::Connection. Use path.walk in Engine::Route

### PR DESCRIPTION
**Pull request by roseundy**
Completely removes the Engine::Connection class. Laying a tile now longer needs to call connect!, instead Engine::Route uses Path::walk to discover connections to the next node (now named chains).

Performance fixes to 1860 and 18Mag. Slight change to 1861 and 1867 to deal with the loss of Hex.connections.

Minor changes to 1822, 1849, 1873, UI::RouteSelector, UI::MapControls to replace call to Routes::connections with Routes::chains, and to rename 'connections' to 'connection_hexes' where appropriate.

### CODE REVIEW

File: lib/engine/game/g_1861/game.rb [hunk 1]
```diff
@@ -868,15 +868,26 @@ def home_token_locations(corporation)
             hex.tile.cities.any? { |city| city.tokenable?(corporation, free: true) && city.tokens.none? }
           end
 
-          # @todo: this may need optimizing when changing connections for loading.
-          unconnected = open_locations.select { |hex| hex.connections.none? }
+          unconnected = open_locations.reject { |hex| connected_hex?(hex) }
```
Comment (tobymao) [inline lib/engine/game/g_1861/game.rb:871]: delete connected_hex? change it to

def unconnected_hexes 
unconnected.empty?

File: lib/engine/route.rb [hunk 2]
```diff
@@ -269,36 +293,77 @@ def corporation
...
+          uniq_paths << path if !junction_map[path.a] && !junction_map[path.b]
+          junction_map[path.a] = true if path.a.junction?
+          junction_map[path.b] = true if path.b.junction?
+        end
+
+        uniq_paths.map! { |path| path.hex.id }
+      end
+    end
+
+    def chain_hexes(paths)
```
Comment (tobymao) [inline lib/engine/route.rb:325]: delete this method, put it inline

File: lib/engine/route.rb [hunk 3]
```diff
@@ -269,36 +293,77 @@ def corporation
...
 
+    def find_matching_chains(hex_ids)
+      start_hex = @game.hex_by_id(hex_ids.first.split.first)
+      end_hex = @game.hex_by_id(hex_ids.last.split.first)
+      matching = []
+      start_hex.tile.nodes.each do |start_node|
+        end_hex.tile.nodes.each do |end_node|
+          next if start_node == end_node
+
+          chains = get_node_chains(start_node, end_node)
```
Comment (tobymao) [inline lib/engine/route.rb:358]: get_node_chains(...)each do

File: lib/engine/route.rb [hunk 4]
```diff
@@ -59,56 +60,79 @@ def tail
...
-      nodes = [node, other]
+      get_node_chains(node, other).select { |c| (c[:paths] & other_paths).empty? }
+    end
 
-      node.hex.all_connections.select do |c|
-        (c.nodes & nodes).size == 2 && (c.paths & other_paths).empty?
-      end
+    # walk paths from start_node, and only keep connections that end at end_node
+    def get_node_chains(start_node, end_node)
+      @node_chains[[start_node, end_node]] ||=
```
Comment (tobymao) [inline lib/engine/route.rb:91]: you never need to clear the cache?

File: lib/engine/game/g_1861/game.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/game/g_1861/game.rb:871]: I'm sorry, I don't understand what you want here.

File: lib/engine/route.rb [hunk 5]
```diff
@@ -269,36 +293,77 @@ def corporation
...
 
     private
 
+    def chain_id(paths)
+      # deal with ambiguous intra-tile path
+      if paths.one? && paths[0].tile.ambiguous_connection?
+        node0, node1 = paths[0].nodes.map(&:index).sort
+        ["#{paths[0].hex.id} #{node0}.#{node1}"]
+      else
+        uniq_paths = []
```
Comment (tobymao) [inline lib/engine/route.rb:311]: hex_ids = []

File: lib/engine/route.rb [hunk 6]
```diff
@@ -269,36 +293,77 @@ def corporation
...
+        junction_map = {}
+
+        # skip over paths that have a junction we've already seen
+        paths.each do |path|
+          uniq_paths << path if !junction_map[path.a] && !junction_map[path.b]
```
Comment (tobymao) [inline lib/engine/route.rb:316]: hex_ids << path.hex.id

File: lib/engine/route.rb [hunk 7]
```diff
@@ -269,36 +293,77 @@ def corporation
...
+          junction_map[path.a] = true if path.a.junction?
+          junction_map[path.b] = true if path.b.junction?
+        end
+
+        uniq_paths.map! { |path| path.hex.id }
```
Comment (tobymao) [inline lib/engine/route.rb:321]: don't need this

File: lib/engine/route.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/route.rb:91]: I don't think so, since it should never change based on user input.

File: lib/engine/route.rb [hunk 8]
```diff
@@ -59,56 +60,79 @@ def tail
       connection_data[-1]
     end
 
+    def chains
+      connection_data&.map { |c| c[:chain] }
+    end
+
     def connections
```
Comment (tobymao) [inline lib/engine/route.rb:67]: is this method used anymore?

File: lib/engine/route.rb [hunk 9]
```diff
@@ -269,36 +293,77 @@ def corporation
...
-        connections_b.each do |b|
-          next if (middle = (a.nodes & b.nodes)).empty?
-          next if (b.paths & a.paths).any?
+    def find_pairwise_chain(chains_a, chains_b, other_paths)
+      chains_a = chains_a.select { |a| (a[:paths] & other_paths).empty? }
+      chains_b = chains_b.select { |b| (b[:paths] & other_paths).empty? }
+      chains_a.each do |a|
+        chains_b.each do |b|
+          next if (middle = (a[:nodes] & b[:nodes])).empty?
+          next if (b[:paths] & a[:paths]).any?
```
Comment (tobymao) [inline lib/engine/route.rb:339]: unless empty?

File: lib/engine/route.rb (hunk 8 above)
Comment (roseundy) [reply lib/engine/route.rb:67]: Yes, by the UI.

File: lib/engine/route.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/route.rb:325]: OK

File: lib/engine/route.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/route.rb:358]: OK

File: lib/engine/route.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/route.rb:311]: OK

File: lib/engine/route.rb (hunk 6 above)
Comment (roseundy) [reply lib/engine/route.rb:316]: OK

File: lib/engine/route.rb (hunk 7 above)
Comment (roseundy) [reply lib/engine/route.rb:321]: OK

File: lib/engine/route.rb (hunk 9 above)
Comment (roseundy) [reply lib/engine/route.rb:339]: OK

File: lib/engine/game/g_1861/game.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/game/g_1861/game.rb:871]: OK. I think I've got it.

Comment (roseundy) [reply lib/engine/game/g_1861/game.rb:871]: I could simplify this further, but I want to re-use the same unconnected_hexes method for 61 and 67

File: lib/engine/route.rb (hunk 8 above)
Comment (roseundy) [reply lib/engine/route.rb:67]: Actually, replaced all calls to this with chains

**Review by tobymao — APPROVED** [review]

**Pull request merged**